Enable test cases for feature `LE security mode 3` (#1454)
auto-pts/auto-pts: doc/btp_gap.txt

                In case of an error, the error response will be returned.

        Opcode 0x2b - Synchronize to a BIG

                Controller Index:       <controller id>
                Command parameters:     Address_Type (1 octet)
                                        Address (6 octets)
                                        SID (1 octet)
                                        Num_Bis (1 octet)
                                        Bis_Bitfield (1 octet)
                                        Mse (4 octets)
                                        Sync_Timeout (2 octets)
                                        Encryption (1 octet)

>                                         Broadcast_Code (16 octets)

Author: Updated.

                In case of an error, the error response will be returned.

        Opcode 0x2b - Synchronize to a BIG

                Controller Index:       <controller id>
                Command parameters:     Address_Type (1 octet)
                                        Address (6 octets)
                                        SID (1 octet)

>                                         Num_Bis (1 octet)

Author: Updated.

>                                         Bis_Bitfield (1 octet)

Author: Updated.

                                        Mse (4 octets)
                                        Sync_Timeout (2 octets)

>                                         Encryption (1 octet)
>                                         Broadcast_Code (0 octet or 16 octets)

Reviewer: ```suggestion
                                        Broadcast_Code (0 octet or 16 octets)
```

Since we can use the size of the command to determine whether a broadcast code is supplied or not, I suggest that we remove the encryption field

Author: Uh huh, I feel like the readability has further decreased if parameter Encryption is omitted.

                Return Parameters:      <none>

                This command is used to synchronize to a BIG. Before executing this command, the
                IUT should be synced to a PADV. It means the `Periodic Sync Established event`
                (opcode 0x8d) should be received before executing this command.

                Available Encryption:
                0 - The BIS will be unencrypted. The Broadcast_Code parameter should be omitted.
                1 - The BIS will be encrypted. The Broadcast_Code parameter should be existing and
                    the length should be 16 octets.

                In case of an error, the error response will be returned.
                In case of a success, the IUT continues processing the command asynchronously.

        Opcode 0x2c - Create BIG

                Controller Index:       <controller id>

>                 Command parameters:     Id (1 octet)

Author: It is a ID corresponding to a specific ID of extended advertising.

>                                         Num_Bis (1 octet)

Author: Updated.

                                        Interval (4 octets)
                                        Latency (2 octets)
                                        Packing (1 octet)
                                        Framing (1 octet)

>                                         Encryption (1 octet)
>                                         Broadcast_Code (0 octet or 16 octets)

Reviewer: ```suggestion
                                        Broadcast_Code (0 octet or 16 octets)
```

                Return Parameters:      <none>

                This command is used to create a BIG.

                Available Encryption:
                0 - The BIS will be unencrypted. The Broadcast_Code parameter should be omitted.
                1 - The BIS will be encrypted. The Broadcast_Code parameter should be existing and
                    the length should be 16 octets.

                In case of an error, the error response will be returned.

        Opcode 0x2d - BIS Broadcast

                Controller Index:       <controller id>

>                 Command parameters:     Bis_Id (1 octet)

Author: It is the value reported by "Opcode 0x94 - BIS Data Path Setup Event".

>                 Command parameters:     Bis_Id (1 octet)

Author: Updated.

                IUT should be synced to a PADV. It means the `Periodic Sync Established event`
                (opcode 0x8d) should be received before executing this command.

                Possible values for the BIS_Bitfield parameter are a bit-wise OR of the bits 0-31.
                Such as, bit 0 represents BIS index 1, bit 1 represents BIS index 2, and so on.
                And the number of bits set by the BIS_Bitfield parameter should be consistent with
                the Num_BIS parameter.

                Available Encryption:
                0 - The BIS will be unencrypted. The Broadcast_Code parameter should be omitted.
                1 - The BIS will be encrypted. The Broadcast_Code parameter should be existing and
                    the length should be 16 octets.

                In case of an error, the error response will be returned.
                In case of a success, the IUT continues processing the command asynchronously.

                After the command is processed, the IUT shall setup the Data Path for all the BIS,
                and generate a "BIS Data Path Setup Event" for each BIS.

>         Opcode 0x2c - Create BIG

Author: I consider that setup data path is a necessary step after creating big. So when big is created, the setup data path event of bis will notify.

                                        Packing (1 octet)
                                        Framing (1 octet)
                                        Encryption (1 octet)
                                        Broadcast_Code (0 octet or 16 octets)
                Return Parameters:      <none>

                This command is used to create a BIG. The parameter ID is corresponding to a
                specific periodic advertising.

                Available Encryption:
                0 - The BIS will be unencrypted. The Broadcast_Code parameter should be omitted.
                1 - The BIS will be encrypted. The Broadcast_Code parameter should be existing and
                    the length should be 16 octets.

                In case of an error, the error response will be returned.
                In case of a success, the IUT continues processing the command asynchronously.

                After the command is processed, the IUT shall setup the Data Path for all the BIS,
                and generate a "BIS Data Path Setup Event" for each BIS.

        Opcode 0x2d - BIS Broadcast

                Controller Index:       <controller id>

>                 Command parameters:     BIS_ID (1 octet)

Reviewer: Right, but if sync to e.g. BIS index 1, 3 and 5, how do I map those BIS_IDs to BIS Indexes?

>                 Command parameters:     BIS_ID (1 octet)

Author: The BIS ID depends on the IUT. From auto-pts side, the report the BIS ID of "Opcode 0x94 - BIS Data Path Setup Event" is the parameter of this command.

                This command is used to synchronize to a BIG. Before executing this command, the
                IUT should be synced to a PADV. It means the `Periodic Sync Established event`
                (opcode 0x8d) should be received before executing this command.

                Possible values for the BIS_Bitfield parameter are a bit-wise OR of the bits 0~31.
                Such as, bit 0 represents BIS index 1, bit 1 represents BIS index 2, and so on.
                And the number of bits set by the BIS_Bitfield parameter should be consistent with
                the Num_BIS parameter.

                Available Encryption:
                0 - The BIS will be unencrypted. The Broadcast_Code parameter should be omitted.
                1 - The BIS will be encrypted. The Broadcast_Code parameter should be existing and
                    the length should be 16 octets.

                In case of an error, the error response will be returned.
                In case of a success, the IUT continues processing the command asynchronously.

                After the command is processed, the "BIS Data Path Setup Event" will be notified

>                 subsequently.

Author: Udpated.

                In case of a success, the IUT continues processing the command asynchronously.

                After the command is processed, the IUT shall setup the Data Path for all the BIS,
                and generate a "BIS Data Path Setup Event" for each BIS.

        Opcode 0x2c - Create BIG

                Controller Index:       <controller id>
                Command parameters:     ID (1 octet)
                                        Num_BIS (1 octet)
                                        Interval (4 octets)
                                        Latency (2 octets)
                                        RTN (1 octet)
                                        PHY (1 octet)
                                        Packing (1 octet)
                                        Framing (1 octet)
                                        Encryption (1 octet)
                                        Broadcast_Code (0 octet or 16 octets)
                Return Parameters:      <none>

                This command is used to create a BIG. The parameter ID is corresponding to a
                specific periodic advertising.

>                 Possible values for the PHY parameter are a bit-wise or of the following bits:
>                         0 = PHY 1M
>                         1 = PHY 2M
>                         2 = PHY Coded

Reviewer: ```suggestion
                Possible values for the PHY parameter are a bitfield or of the following bits:
                        0 = PHY 1M
                        1 = PHY 2M
                        2 = PHY Coded
```

What does it mean if you provide e.g. `0b111`? Why can more than 1 bit be set?

Reviewer: Right, but I think that @sjanc 's plan is to not lean too much on the HCI spec, so we need to specify this in a BTP context

                Available Encryption:

>                 0 - The BIS will be unencrypted. The Broadcast_Code parameter should be omitted.
>                 1 - The BIS will be encrypted. The Broadcast_Code parameter should be existing and
>                     the length should be 16 octets.

Reviewer: ```suggestion
                        0 - The BIS will be unencrypted. The Broadcast_Code parameter should be omitted.
                        1 - The BIS will be encrypted. The Broadcast_Code parameter should be existing and
                            the length should be 16 octets.
```

To be more consistent with other definitions

                This event can be sent after IUT received encryption change event was received.
                The parameter `Enabled` is the flag whether the link is encrypted.

        Opcode 0x92 - BIG Sync Established event

                Controller Index:       <controller id>
                Event parameters:       Address_Type (1 octet)

>                                         Address (6 octets)

Author: Well. The information cannot be got directly in the parameters of callback `struct bt_iso_big_cb::started`. It could be read through the function `bt_iso_chan_get_info()`.

>                                         Address (6 octets)

Author: Well. All information is not required now. But I think it could be reported. I will add it.

>                                         Address (6 octets)

Author: Added.

                This event can be sent after IUT received the BIG sync established event was
                received.

        Opcode 0x93 - BIS Setup Event

                Controller Index:       <controller id>
                Event parameters:       Address_Type (1 octet)
                                        Address (6 octets)
                                        Bis_Id (1 octet)

>                 This event can be sent after IUT received the BIS setup event was received.

Author: The `BIS Setup Event` will be notified if the HCI command `LE Setup ISO Data Path` has been sent.

>                 This event can be sent after IUT received the BIS setup event was received.

Author: Update the event name to `BIS Data Path Setup Event`.

                Event parameters:       Address_Type (1 octet)
                                        Address (6 octets)
                                        Latency (4 octets)
                                        Nse (1 octet)
                                        Bn (1 octet)
                                        Pto (4 octets)
                                        Irc (1 octet)
                                        Max_Pdu (2 octets)
                                        ISO_interval (2 octets)

                This event can be sent after IUT received the BIG sync established event was
                received.

        Opcode 0x93 - BIG Sync Lost event

                Controller Index:       <controller id>
                Event parameters:       Address_Type (1 octet)
                                        Address (6 octets)

                This event can be sent after IUT received the BIG sync lost event was received.

        Opcode 0x94 - BIS Data Path Setup Event

                Controller Index:       <controller id>
                Event parameters:       Address_Type (1 octet)
                                        Address (6 octets)

>                                         Bis_Id (1 octet)

Reviewer: > It is a logical concept. It is a tag corresponding to a specific BIS. I think it could be a ID or index.

Not sure I follow. It is either an ID or an index, and that needs to be explicit, because those 2 terms are not the same. You are currently using this as a key to a dictionary, so it's being used as an ID and not an index, which is fine :) Just wanted to make sure

>                                         Bis_Id (1 octet)

Author: Yes. I think it is an ID corresponding to a specific BIS.

                This event can be sent after IUT has been setup the data path for the specific BIS.

        Opcode 0x95 - BIS Stream Received Event

                Controller Index:       <controller id>
                Event parameters:       Address_Type (1 octet)
                                        Address (6 octets)

>                                         Bis_Id (1 octet)

Author: It is the value reported by "Opcode 0x94 - BIS Data Path Setup Event".

>                                         Bis_Id (1 octet)

Author: Updated.

                                        Address (6 octets)
                                        BIS_ID (1 octet)
                                        Flags (1 octet)
                                        Timestamp (4 octets)
                                        Sequence_Number (2 octets)
                                        Data Length (1 octet)
                                        Data (variable)

                This event can be sent after IUT received the stream data from the specific BIS.

                Available Flags:
                        Bit 0: The packet is valid
                        Bit 1: The packet may possibly contain errors
                        Bit 2: The Packet was lost
                        Bit 3: The Timestamp parameter is valid

        Opcode 0x96 - Periodic BIGInfo Event

                Controller Index:       <controller id>
                Event parameters:       Address_Type (1 octet)
                                        Address (6 octets)
                                        Sync Handle (2 octets)
                                        SID (1 octet)
                                        Num_BIS (1 octet)

>                                         Encryption (1 octet)

Reviewer: May be, but we need to be future safe. This is a protocol definition and it's non-trivial to add things later (which is an issue we've seen quite a lot with BTP)

>                                         Encryption (1 octet)

Author: Added.

                                        BIS_ID (1 octet)
                                        Flags (1 octet)
                                        Timestamp (4 octets)
                                        Sequence_Number (2 octets)
                                        Data Length (1 octet)
                                        Data (variable)

                This event can be sent after IUT received the stream data from the specific BIS.

                Available Flags:
                        Bit 0: The packet is valid
                        Bit 1: The packet may possibly contain errors
                        Bit 2: The Packet was lost
                        Bit 3: The Timestamp parameter is valid

        Opcode 0x96 - Periodic BIGInfo Event

                Controller Index:       <controller id>
                Event parameters:       Address_Type (1 octet)
                                        Address (6 octets)
                                        Sync Handle (2 octets)
                                        SID (1 octet)
                                        Num_BIS (1 octet)
                                        NSE (1 octet)

>                                         ISO_Interval (2 octets)

Reviewer: What is the unit of ISO_Interval?

Author: Updated the doc to explain it.

                                        BN (1 octet)
                                        PTO (1 octet)
                                        IRC (1 octet)
                                        Max_PDU (2 octets)

>                                         SDU_Interval (4 octets)

Reviewer: What is the unit of SDU_Interval?

                                        Max_SDU (2 octets)
                                        PHY (1 octet)
                                        Framing (1 octet)

>                                         Encryption (1 octet)

Reviewer: I realize this is a general issue with the BTP definitions, but we really should define the allowed value range (if not the full range of the value) and unit for each field for each event and command. 

For example SDU_Interval is defined to just be 4 octets, but it doesn't mention what the unit is, or if 0xFFFFFFFF is a valid value. 

Additionally we really should add verification of each field for each event and command as well. I know that we don't really do this much in BTP, but it has caused issues over time in my experience, and we should work towards maturing the BTP definitions. cc @sjanc

                This event can be sent after IUT received the periodic BIGInfo Advertising Report
                on established sync.

>                 Available Encryption:

Reviewer: You provide the possible values of the Encryption field here, but not e.g. the Framing or PHY fields

>                 Available Encryption:
>                 0 - The BIS is unencrypted.
>                 1 - The BIS is encrypted.

Reviewer: ```suggestion
                Available Encryption:
                        0 - The BIS is unencrypted.
                        1 - The BIS is encrypted.
```

Reviewer: To be more consistent with other definitions
